syft/cpe/by_source_then_specificity.go

```diff
@@ -0,0 +1,40 @@
+package cpe
+
+import "sort"
+
+type BySourceThenSpecificity []SourcedCPE
+
+func (b BySourceThenSpecificity) Len() int {
+	return len(b)
+}
+
+func (b BySourceThenSpecificity) Less(i, j int) bool {
+	sourceOrder := map[Source]int{
+		NVDDictionaryLookupSource: 1,
+		DeclaredSource:            2,
+		GeneratedSource:           3,
+	}
+
+	// Function to get the rank of a source
+	getRank := func(source Source) int {
+		if rank, exists := sourceOrder[source]; exists {
+			return rank
+		}
+		return 4 // Sourced we don't know about can't be assigned special priority, so
+		// are considered ties.
+	}
+	iSource := b[i].Source
+	jSource := b[j].Source
+	rankI, rankJ := getRank(iSource), getRank(jSource)
+	if rankI != rankJ {
+		return rankI < rankJ
+	}
+
+	return weightedCountForSpecifiedFields(b[i].CPE) < weightedCountForSpecifiedFields(b[j].CPE)
```

Comment:
FIXME: this comparison is backwards; more specific CPEs should be first.

Reply:
Also, extract the Less from BySpecificity into a helper method and then call it here to break ties.